Invite UI

Nikunjsharma0 opened the pull request:
invite via contact ui created from android / desktop

Nikunjsharma0 committed:
invite via contact ui created for android/desktop

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/domain/invite/ParseInviteUseCase.kt:
```diff
@@ -0,0 +1,45 @@
+        phoneValidatorUseCase: PhoneNumValidatorUseCase
+    ): List<InviteViaContactChip> {
+        phoneNumValidatorUseCase = phoneValidatorUseCase
+        val parts = text.split(",\\s*".toRegex())
```
Better to use .split(",").map { it.trim() } . Regex not needed here - and it can be tricky to predict exactly what a regex will do in any given edge case.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/domain/invite/ParseInviteUseCase.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private fun isValidEmail(email: String): Boolean {
+        ValidateEmailUseCase().invoke(email)
```
ValidateEmailUseCase should also be a constructor parameter.

mikedawson commented on core/src/jvmTest/kotlin/com/ustadmobile/core/domain/invite/ParseInviteUseCaseTest.kt:
This is a good start. Please add scenarios for invalid input.

mikedawson commented on lib-ui-compose/build.gradle.kts:
```diff
@@ -122,6 +122,9 @@ android {
         jvmToolchain(17)
     }
 }
+//dependencies {
```
Please remove this. Generally we should not have commented out code in commits. If we need to refer to what the code was at some point before, then that is what Git is for. The only exception is when there is specific code that we would restore once some known issue is resolved.
Whenever we have commented out code, there should be a comment before it explaining why it is commented out instead of just removed.

mikedawson commented on lib-ui-compose/src/androidMain/kotlin/com/ustadmobile/libuicompose/components/UstadContactPickButton.kt:
```diff
@@ -0,0 +1,158 @@
+            if (contactUri != null) {
+                val projection = arrayOf(ContactsContract.CommonDataKinds.Phone.NUMBER)
+                val cursor = context.contentResolver.query(contactUri, projection, null, null, null)
+                CoroutineScope(Dispatchers.IO).launch {
```
Instead of creating a new coroutine scope - use the coroutine scope that is provided from compose so that any task would be cancelled automatically if the user navigates away.

eg.

val scope = rememberCoroutineScope()

then to run something off the main thread
```
scope.launch {
   withContext(Dispatchers.IO) {
       //do something - will be cancelled automatically if user went away
   } 
}
```
scope.launch(Dispatchers.IO) might also work

mikedawson commented on lib-ui-compose/src/androidMain/kotlin/com/ustadmobile/libuicompose/components/UstadContactPickButton.kt:
```diff
@@ -0,0 +1,158 @@
+                }
+
+            } else {
+                onContactPicked(null)
```
If the user didn't select anything; then we don't need onContactPicked(null). 

If they did... but there was some kind of error (e.g. it.moveToFirst() doesn't have anything) then we should have a different function e.g. onContactError(t: Throwable?, message: String?). That can then go to the viewmodel, which can show an error using showSnackBar

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/clazz/inviteViaContact/InviteViaContactScreen.kt:
```diff
@@ -0,0 +1,147 @@
+) {
+    val state = rememberChipTextFieldState<AvatarChip>()
+    val scrollState = rememberScrollState()
+    Column(
```
Please use UstadVerticalScrollColumn - it takes care of showing the scrollbar itself on Desktop.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/clazz/inviteViaContact/InviteViaContactScreen.kt:
```diff
@@ -0,0 +1,147 @@
+            },
+
+            chipStyle = ChipTextFieldDefaults.chipStyle(shape = RoundedCornerShape(20.dp)),
+            textStyle = TextStyle(fontSize = 13.sp),
```
Please use a MaterialTheme for text style.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/clazz/inviteViaContact/InviteViaContactScreen.kt:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    LaunchedEffect(uiState.chips) {
+        uiState.chips?.let { validChip ->
```
Bad variable name: you don't know its a valid chip at this line.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/clazz/inviteViaContact/InviteViaContactScreen.kt:
```diff
@@ -0,0 +1,147 @@
+        )
+    }
+
+    LaunchedEffect(uiState.chips) {
```
I don't think this function will work as intended. Every time the uiState chips change it will add it again

e.g. user adds chip A, goes to view model, comes back on ui State, gets added to chip state [uistate = [A], chip state = [A].
user adds chip B, goes to view model, come back on ui state as [A,B], both get added, now chip state [ A, A, B]

mikedawson replied:
This needs to check the material chips state to see if the given chip is already there before adding it.

mikedawson replied:
Also I don't see code handling the removal.

Nikunjsharma0 replied:
we dont neet to worry about removal of chips because when we remove any chip this state = rememberChipTextFieldState<AvatarChip>()  will remove chip on it own

mikedawson replied:
Yes we do need to worry about removal: when the user actually goes and clicks invite, then the viewmodel needs to know (otherwise the invite request will include items removed by the user).

Another reason that state needs to be in the ViewModel: we use the saved state handle (in the view model) e.g. if the app is destroyed when the user leaves due to inactivity, and then returns, we should restore the state.

mikedawson reviewed (changes requested):
Generally

Nikunjsharma0 committed:
chip text field changes

Nikunjsharma0 committed:
clazzinvite added to dp, processinviteUsecase class create to send invite

Nikunjsharma0 committed:
route created for clazzinviteviacontact

Nikunjsharma0 committed:
ContactToServerUseCase created

Nikunjsharma0 committed:
api response getting,
ui error cases handled

Nikunjsharma0 committed:
twilio added for sms
and invite via username also added

Nikunjsharma0 committed:
assetlink json created

Nikunjsharma0 committed:
Merge branch 'refs/heads/primary' into invite_users

# Conflicts:
#	app-android/src/main/AndroidManifest.xml
#	app-ktor-server/src/main/kotlin/com/ustadmobile/lib/rest/UmRestApplication.kt
#	core/src/commonMain/resources/MR/base/strings.xml
#	lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/UmAppDatabase.kt
#	lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/ext/UmAppDatabaseMigrations.kt

Nikunjsharma0 committed:
data base version changes

mikedawson committed:
Remove hard coded assetlinks.json

mikedawson committed:
Add well-known to gitignore

Nikunjsharma0 committed:
link json verified

Nikunjsharma0 committed:
invite accept flow completed

Nikunjsharma0 committed:
Merge branch 'refs/heads/primary' into invite_users

# Conflicts:
#	core/src/commonMain/resources/MR/base/strings.xml

Nikunjsharma0 committed:
in web ui chip text fields connected with viewmodel
and invite accept and decline screen created.

in ClazzInvite invitestatus fieldadded

Nikunjsharma0 committed:
withDoorTransactionAsync added in ProcessInviteUseCase

mikedawson commented:
This pull request is now obsolete since it has been merged into dev-simplified-onboarding, and will be merged to primary through that.
